test/e2e/util/fixture.go

```diff
@@ -0,0 +1,255 @@
+
+	// Try repeatedly to destroy the cluster gracefully. For each failure, dump
+	// the current cluster to help debug teardown lifecycle issues.
+	t.Run("DestroyCluster", func(t *testing.T) {
+		t.Logf("Waiting for cluster to be destroyed. Namespace: %s, name: %s", hc.Namespace, hc.Name)
+		testName := strings.ReplaceAll(t.Name(), "/", "_")
+		destroyAttempt := 1
+		err := wait.PollImmediateUntil(5*time.Second, func() (bool, error) {
+			err := destroyCluster(ctx, hc, opts)
+			if err != nil {
+				t.Logf("Failed to destroy cluster, will retry: %v", err)
+				if len(artifactDir) > 0 {
+					dumpPath := filepath.Join(artifactDir, testName, fmt.Sprintf("destroy-hostedcluster-%d", destroyAttempt))
+					err := dumpCluster(ctx, t, hc, opts, dumpPath)
+					if err != nil {
+						t.Logf("Failed to dump cluster during destroy; this is nonfatal: %v", err)
+					}
+				}
+				return false, nil
+			}
+			return true, nil
+		}, ctx.Done())
+		if err != nil {
+			t.Errorf("Failed to destroy cluster: %v", err)
+		} else {
+			t.Logf("Destroyed cluster. Namespace: %s, name: %s", hc.Namespace, hc.Name)
+		}
+	})
+
+	// Finally, delete the test namespace containing the HostedCluster/NodePool
+	// resources.
+	//
+	// If the cluster was successfully destroyed and finalized, any further delay
+	// in cleaning up the test namespace could be indicative of a resource
+	// finalization bug. Give this namespace teardown a reasonable time to
+	// complete and then dump resources to help debug.
+	t.Run("DeleteTestNamespace", func(t *testing.T) {
+		deleteTimeout, cancel := context.WithTimeout(ctx, 1*time.Minute)
+		defer cancel()
+		err := DeleteNamespace(t, deleteTimeout, client, hc.Name)
+		if err != nil {
+			t.Errorf("Failed to delete test namespace: %v", err)
+			if len(artifactDir) > 0 {
+				err := dumpCluster(ctx, t, hc, opts, artifactDir)
+				if err != nil {
+					t.Errorf("Failed to dump cluster: %v", err)
+				}
+			}
+		}
+	})
+}
+
+// createClusterOpts mutates the cluster creation options according to the
+// cluster's platform as necessary to deal with options the test caller doesn't
+// know or care about in advance.
+//
+// TODO: Mutates the input, instead should use a copy of the input options
+func createClusterOpts(hc *hyperv1.HostedCluster, opts *core.CreateOptions) *core.CreateOptions {
+	opts.Namespace = hc.Namespace
+	opts.Name = hc.Name
+
+	switch hc.Spec.Platform.Type {
+	case hyperv1.AWSPlatform:
+		opts.InfraID = hc.Name
+	}
+
+	return opts
+}
+
+// createCluster calls the correct cluster create CLI function based on the
+// cluster platform.
+func createCluster(ctx context.Context, hc *hyperv1.HostedCluster, opts *core.CreateOptions) error {
+	switch hc.Spec.Platform.Type {
+	case hyperv1.AWSPlatform:
+		return aws.CreateCluster(ctx, opts)
+	case hyperv1.NonePlatform:
+		return none.CreateCluster(ctx, opts)
+	case hyperv1.KubevirtPlatform:
+		return kubevirt.CreateCluster(ctx, opts)
+	default:
+		panic("unsupported platform")
```

Comment:
Nit: You could pass the `testing.T` thorugh and use `t.Fatal` here

Reply:
thinking about this some more, panic would leak the test namespace itself even though there aren't any resources there, so I think that's a reason to handle the error instead of panic

Reply:
I switched these to just return errors, which will result in graceful teardown handling